flink-core/src/test/java/org/apache/flink/testutils/TupleEqualityMatcher.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.testutils;
+
+import org.apache.flink.api.java.tuple.Tuple;
+
+import org.hamcrest.BaseMatcher;
+import org.hamcrest.Description;
+
+import java.util.Arrays;
+
+public class TupleEqualityMatcher extends BaseMatcher<Object> {
+
+	private final Object wanted;
+
+	/**
+	 * This matcher performs similar comparison to {@link org.hamcrest.core.IsEqual}, which resembles
+	 * {@link java.util.Objects#deepEquals(Object, Object)} logic. The only difference here is that {@link Tuple}s
+	 * are treated similarly to arrays.
+	 *
+	 * <p>This means that if we compare two Tuples that contain arrays, those arrays will
+	 * be compared with {@link Arrays#deepEquals(Object[], Object[])} rather than with reference comparison.
+	 *
+	 * @param item expected value
+	 */
+	public static TupleEqualityMatcher deeplyEquals(Object item) {
+		return new TupleEqualityMatcher(item);
+	}
+
+	private TupleEqualityMatcher(Object wanted) {
+		this.wanted = wanted;
+	}
+
+	@Override
+	public boolean matches(Object item) {
+		return deepEquals(item, wanted);
+	}
+
+	@Override
+	public void describeTo(Description description) {
+		description.appendValue(wanted);
+	}
+
+	private static boolean deepEquals(Object o1, Object o2) {
+		if (o1 == o2) {
+			return true;
+		} else if (o1 == null || o2 == null) {
+			return false;
+		} else if (o1.getClass() != o2.getClass()) {
+			return false;
+		} else if (o1 instanceof Tuple && o2 instanceof Tuple) {
+			Tuple tuple1 = (Tuple) o1;
+			Tuple tuple2 = (Tuple) o2;
+			return deepEqualsTuple(tuple1, tuple2);
+		} else {
+			return deepEquals0(o1, o2);
+		}
+	}
+
+	private static boolean deepEquals0(Object e1, Object e2) {
+		assert e1 != null;
+		boolean eq;
+		if (e1 instanceof Object[] && e2 instanceof Object[]) {
```

Comment:
I think this can be replaced with e1.getClass().isArray() 
and the recursively call deepEquals element wise (to catch Tuples of Tuples)
or if this is not needed for the test, then maybe just replace with `Arrays.deepEquals`